src/Compilers/VisualBasic/Portable/Symbols/Source/SourceClonedParameterSymbol.vb

```diff
@@ -149,6 +149,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             End Get
         End Property
 
+        Friend Overrides ReadOnly Property CallerArgumentExpressionParameterIndex As Integer
+            Get
+                Return _originalParam.CallerArgumentExpressionParameterIndex
```

Comment:
>CallerArgumentExpressionParameterIndex [](http://example.com/codeflow?start=38&length=38)

This looks fragile and probably incorrect, at least for delegate EndInvoke method. This implementation assumes that the target parameter is still going to be present and will remain at the same position, which probably can be wrong. The more robust approach is to reinterpret the attribute in the new signature. But before we do this, let's create some tests and figure out if this code path is even reachable in interesting scenarios. Perhaps simply returningh -1 is going to be good enough.

Reply:
@AlekseyTs EndInvoke should only be cloning parameters that are `ref` or `out`:

https://github.com/dotnet/roslyn/blob/22a72038d011eadbaa5039092fbab5f8a4d30650/src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateMethodSymbol.cs#L402-L409

So it is not going to ever clone a caller argument expression parameter. Hence, this path is unreachable for an EndInvoke. However, the implementation is needed for indexers and delegate invoke methods.

Reply:
> So it is not going to ever clone a caller argument expression parameter. 

Why do you think so? Is there a test that demonstrates that?

Reply:
> However, the implementation is needed for indexers and delegate invoke methods.

I see nothing in this type that limits its application to these scenarios. Therefore, the implementation is not robust in my opinion.

Reply:
> Why do you think so?

Per the code above that creates `SourceClonedParameterSymbol` for `EndInvoke`, a parameter is only created if its `RefKind == RefKind.None`.

`CallerArgumentExpressionAttribute` is only allowed for optional parameters (which should have RefKind.None.

Applying the attribute to a ref or out parameter generates a compile-time error:

>  `error CS9006: The CallerArgumentExpressionAttribute may only be applied to parameters with default values`

*These are just my thoughts, I may be missing some interesting cases that you may have.*

Reply:
> CallerArgumentExpressionAttribute is only allowed for optional parameters (which should have RefKind.None.

Is this accurate? One cannot do this in IL or VB?

> Applying the attribute to a ref or out parameter generates a compile-time error:

> error CS9006: The CallerArgumentExpressionAttribute may only be applied to parameters with default values

Why are we talking about C# error, when we are working in VB? What prevents applying this attribute in IL for example? 

This is perfectly valid code:
```
class Test
    Shared Sub M(Optional ByRef x as String = "ddd")
    End Sub
    
    Shared Sub Test()
        M()
    End SUb
End Class
```

Reply:
> I see nothing in this type that limits its application to these scenarios. Therefore, the implementation is not robust in the long term in my opinion.

One way out could be making this type MustInherit, creating specialized NotInheritable derived types for specific scenarios, implementing the API in the derived types only.

Reply:
@AlekseyTs 

Behavioral-wise:

- VB delegates doesn't allow optional parameter, hence, calling CallerArgumentExpressionIndex should never be done for EndInvoke invocations (the binder calls this property only if we're binding an optional parameter).

- For IL, shouldn't we be dealing with metadata symbols, not source symbols? or I'm misunderstanding the concept here?

----

Implementation-wise:

I'm okay with making the type abstract and introducing derived type.

Reply:
> I'm misunderstanding the concept here?

The discussion like that can go forever and will never end because reviewer's concerns can almost never be addressed by gueses, speculations or asumptions. The best way to alleviate reviewer's concerns is to create a test for the scenario in question and demonstrate the behavior, in this case the behavior of the API. Specifically here we are discussing an API behavior, whether there is an error reported or not has very liitle concern to the question what API's behavior is and what it should be. We are not discussing an issue with lowering/code generation. Since the type is not specific to any single particular scenario (a unique reason for creating a clone), each scenario should be tested individually and the tests should demonstrate that the implementation is in fact necessary and correct. Neccessity can be demonstrated by observing a break for a successful scenario when the implementation is removed/changed. For example, you are saying: "the implementation is needed for indexers and delegate invoke methods." Please demonstrate the fact with unit tests.  

> I'm okay with making the type abstract and introducing derived type.

I would prefer that we will create more specialized derived types. This way one can reason when instances are created and how they are used, can confirm that the code path is tested. Today anyone can create an instance of ```SourceClonedParameterSymbol``` for whatever reason and "take" an implementation of this API, whether appropriate or not, with it without even be aware of the possible impact. I think we should prevent this from happening.